[Merged by Bors] - feat(Probability/Moments): add lemmas about moment generating functions

Vilin97 opened the pull request:
Add `mgf_ident_distrib`, `mgf_sum_iid`.
These are some of the supporting lemmas we have proved on our way to formalizing the central limit theorem.

Co-authored-by: Siyuan Ge <jamesgsy@uw.edu>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -282,6 +287,27 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
   · rw [h_indep.mgf_sum h_meas]
   · exact (mgf_pos (h_int j hj)).ne'
 
+theorem mgf_ident_distrib (X Y : Ω → ℝ) (hident : IdentDistrib X Y μ μ) :
```
Instead of having `X` and `Y` defined on the same space, could you have `X` defined on Omega and `Y` defined on Omega'?

Vilin97 replied:
Thank you, I used your suggestion.

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -282,6 +287,27 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
   · exact (mgf_pos (h_int j hj)).ne'
 
+theorem mgf_ident_distrib (X Y : Ω → ℝ) (hident : IdentDistrib X Y μ μ) :
+  mgf X μ t = mgf Y μ t := by
```
We have quite rigid indentation rules in mathlib (it may look cumbersome at first, but in the long run it helps to have a coherent library). So, except for the first one, the lines of the statement should be indented by 4 spaces, while the lines of proofs should be indented by 2 spaces (plus additional indentation for subgoals). Here, you do the opposite :-)

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -282,6 +287,27 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    rw [compX, compY]
+    exact IdentDistrib.comp hident (Measurable.exp (measurable_const_mul t))
+
+theorem mgf_sum_iid
+{X : ι → Ω → ℝ}
+(h_meas : ∀ (i : ι), Measurable (X i))
+(h_indep : ProbabilityTheory.iIndepFun (fun _ => inferInstance) X μ)
+(hident : ∀ (i j : ι), ProbabilityTheory.IdentDistrib (X i) (X j) μ μ)
+(s : Finset ι) (j : ι)
+: ∀ t : ℝ, mgf (∑ i ∈ s, X i) μ t = (mgf (X j) μ (t)) ^ s.card := by
+    intro t
+    rw [iIndepFun.mgf_sum h_indep h_meas]
+    exact Finset.prod_eq_pow_card (fun i si => mgf_ident_distrib (X i) (X j) (hident i j))
```
Indentation is also not mathlib-like here. Let me show you how it should be:
```suggestion
theorem mgf_sum_iid
    {X : ι → Ω → ℝ}
    (h_meas : ∀ (i : ι), Measurable (X i))
    (h_indep : ProbabilityTheory.iIndepFun (fun _ => inferInstance) X μ)
    (hident : ∀ (i j : ι), ProbabilityTheory.IdentDistrib (X i) (X j) μ μ)
    (s : Finset ι) (j : ι) : 
    ∀ t : ℝ, mgf (∑ i ∈ s, X i) μ t = (mgf (X j) μ (t)) ^ s.card := by
  intro t
  rw [iIndepFun.mgf_sum h_indep h_meas]
  exact Finset.prod_eq_pow_card (fun i si => mgf_ident_distrib (X i) (X j) (hident i j))
```
(possibly regrouping some variables on common lines)

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -331,6 +357,39 @@ theorem measure_le_le_exp_cgf [IsFiniteMeasure μ] (ε : ℝ) (ht : t ≤ 0)
+open Measurable
+
+theorem mgf_gaussian
+(X: Ω → ℝ) (hXm: Measurable X) (hX: (Measure.map X μ) = (gaussianReal 0 1)) (t : ℝ) :
```
Can you also fix the indentation of this one? Also, I'd move it to the files on gaussians, to avoid putting more imports in the current file (which should remain one of the basic files in probability). Maybe also give the statement for non-centered non-normalized gaussians?

Vilin97 commented on Mathlib/Probability/Moments.lean:
```diff
@@ -184,6 +185,9 @@ theorem mgf_neg : mgf (-X) μ t = mgf X μ (-t) := by simp_rw [mgf, Pi.neg_apply
 
 theorem cgf_neg : cgf (-X) μ t = cgf X μ (-t) := by simp_rw [cgf, mgf_neg]
 
+theorem mgf_smul_left (α : ℝ) : mgf (α • X) μ t = mgf X μ (α * t) := by
+  simp_rw [mgf, Pi.smul_apply, smul_eq_mul, mul_comm α t, mul_assoc]
+
```
I am confused why this is flagged as a new change given https://github.com/leanprover-community/mathlib4/pull/19798...

RemyDegenne replied:
If you merge master into the branch of this PR it should go away.

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,28 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
   · exact (mgf_pos (h_int j hj)).ne'
 
+theorem mgf_ident_distrib
+    (X : Ω → ℝ) {Ω' : Type u_1} {m' : MeasurableSpace Ω'} {μ' : Measure Ω'} (X' : Ω' → ℝ)
```
```suggestion
    (X : Ω → ℝ) {Ω' : Type*} {m' : MeasurableSpace Ω'} {μ' : Measure Ω'} (X' : Ω' → ℝ)
```
This is more robust.

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,28 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    (X : Ω → ℝ) {Ω' : Type u_1} {m' : MeasurableSpace Ω'} {μ' : Measure Ω'} (X' : Ω' → ℝ)
+    (hident : IdentDistrib X X' μ μ') (t : ℝ) :
+    mgf X μ t = mgf X' μ' t := by
+  rw [mgf, mgf]
+  apply IdentDistrib.integral_eq
+  let u := fun x => Real.exp (t * x)
+  have compX : (fun ω ↦ Real.exp (t * X ω)) = u ∘ X := rfl
+  have compX' : (fun ω ↦ Real.exp (t * X' ω)) = u ∘ X' := rfl
+  rw [compX, compX']
+  exact IdentDistrib.comp hident (Measurable.exp (measurable_const_mul t))
```
You're telling Lean to rewrite things that it already knows are the same (as you prove them by `rfl`). This means that this is not really necessary, and that you can just skip all these steps, to get:
```suggestion
  apply IdentDistrib.integral_eq
  apply IdentDistrib.comp hident (Measurable.exp (measurable_const_mul t))
```

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,28 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    {X : ι → Ω → ℝ}
+    (h_meas : ∀ (i : ι), Measurable (X i))
+    (h_indep : ProbabilityTheory.iIndepFun (fun _ => inferInstance) X μ)
+    (hident : ∀ (i j : ι), ProbabilityTheory.IdentDistrib (X i) (X j) μ μ)
```
could you assume `hident` only for `i, j` in `s`?

Vilin97 replied:
Yes, done.

sgouezel commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,28 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    (h_meas : ∀ (i : ι), Measurable (X i))
+    (h_indep : ProbabilityTheory.iIndepFun (fun _ => inferInstance) X μ)
+    (hident : ∀ (i j : ι), ProbabilityTheory.IdentDistrib (X i) (X j) μ μ)
+    (s : Finset ι) (j : ι) : ∀ t : ℝ, mgf (∑ i ∈ s, X i) μ t = (mgf (X j) μ (t)) ^ s.card := by
```
```suggestion
    (s : Finset ι) (j : ι) (t : ℝ) : 
    mgf (∑ i ∈ s, X i) μ t = (mgf (X j) μ t) ^ s.card := by
```
Put `t` left of the colon, in this way you don't need to `intro` it in the proof below.

sgouezel reviewed (commented):
I didn't ask that you should remove the result for the Gaussian, only that you move it to the file about the Gaussian. I think it's definitely a good result to have!

Vilin97 commented:
@sgouezel , I just moved the gaussian mgf lemma in another PR: https://github.com/leanprover-community/mathlib4/pull/19896.

I addressed all your comments.

Vilin97 commented:
@sgouezel , friendly ping. Thank you for your attention!

YaelDillies commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,23 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    (hident : ∀ (i j : s), ProbabilityTheory.IdentDistrib (X i) (X j) μ μ)
+    (t : ℝ) : mgf (∑ i ∈ s, X i) μ t = (mgf (X j) μ (t)) ^ s.card := by
+  rw [iIndepFun.mgf_sum h_indep h_meas]
+  exact Finset.prod_eq_pow_card (fun i si => mgf_ident_distrib (X i) (X j) (hident ⟨i,si⟩ j) t)
```
```suggestion
  exact Finset.prod_eq_pow_card fun i si => mgf_ident_distrib (X i) (X j) (hident ⟨i,si⟩ j) t
```

YaelDillies reviewed (commented):
Thank you for the PR! Should be easy to get in

Vilin97 commented:
Thanks, @YaelDillies. I applied your suggestions.

YaelDillies commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,21 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    {X : ι → Ω → ℝ}
+    (h_meas : ∀ (i : ι), Measurable (X i))
+    (h_indep : ProbabilityTheory.iIndepFun (fun _ => inferInstance) X μ)
+    (s : Finset ι) (j : s)
```
It's better to have `j : ι` + `j ∈ s` rather than ` j : s`
```suggestion
    {s : Finset ι} {j : ι} (hj : j ∈ s)
```
Also I think `hj` should be the last proof obligation in this statement, namely it should come just before `(t : ℝ)`

YaelDillies commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,21 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    (h_meas : ∀ (i : ι), Measurable (X i))
+    (h_indep : ProbabilityTheory.iIndepFun (fun _ => inferInstance) X μ)
+    (s : Finset ι) (j : s)
+    (hident : ∀ (i j : s), ProbabilityTheory.IdentDistrib (X i) (X j) μ μ)
```
```suggestion
    (hident : ∀ i ∈ s, ∀ j ∈ s, IdentDistrib (X i) (X j) μ μ)
```

YaelDillies commented on Mathlib/Probability/Moments.lean:
```diff
@@ -4,6 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Rémy Degenne
 -/
 import Mathlib.Probability.Variance
+import Mathlib.Probability.IdentDistrib
```
Please keep imports alphabetical
```suggestion
import Mathlib.Probability.IdentDistrib
import Mathlib.Probability.Variance
```

YaelDillies reviewed (commented):
Thank you for implementing the changes! Here are a few more suggestions

Vilin97 commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,21 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+
+theorem mgf_sum_of_identDistrib
+    {X : ι → Ω → ℝ}
+    (h_meas : ∀ (i : ι), Measurable (X i))
```
But `hident` does not imply `Measurable (X i) μ`, right? And we need it for `mgf_sum`. 
A later refactor might weaken the measurability assumption of `mgf_sum` to `AEMeasurable` or `AEStronglyMeasurable` instead of `Measurable`. But for now, I need `Measurable (X i) μ`.

YaelDillies replied:
Ugh, that's annoying, you're right. Can you leave a TODO then?

YaelDillies commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +286,22 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+    (h_indep : iIndepFun (fun _ => inferInstance) X μ)
+    (hident : ∀ i ∈ s, ∀ j ∈ s, IdentDistrib (X i) (X j) μ μ)
+    (hj : j ∈ s) (t : ℝ) : mgf (∑ i ∈ s, X i) μ t = mgf (X j) μ t ^ #s := by
+  rw [iIndepFun.mgf_sum h_indep h_meas]
```
```suggestion
  rw [h_indep.mgf_sum h_meas]
```

YaelDillies commented on Mathlib/Probability/Moments.lean:
```diff
@@ -261,6 +262,7 @@ theorem iIndepFun.integrable_exp_mul_sum [IsFiniteMeasure μ] {X : ι → Ω →
     refine IndepFun.integrable_exp_mul_add ?_ (h_int i (mem_insert_self _ _)) h_rec
     exact (h_indep.indepFun_finset_sum_of_not_mem h_meas hi_notin_s).symm
 
+-- TODO(vilin97): weaken `h_meas` to `AEMeasurable (X i)` or `AEStronglyMeasurable (X i)` throughout
```
Are you planning on doing this refactor yourself in the near future? If not, you could open an issue and link it here

Vilin97 replied:
I can try, but probably not in the next two weeks. I made an issue: https://github.com/leanprover-community/mathlib4/issues/20367.

kmill replied:
Now that you have an issue number, could you please add it to the comment? Maybe like this?
```suggestion
-- TODO(vilin97): weaken `h_meas` to `AEMeasurable (X i)` or `AEStronglyMeasurable (X i)` throughout
-- https://github.com/leanprover-community/mathlib4/issues/20367
```

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

kmill commented on Mathlib/Probability/Moments.lean:
```diff
@@ -285,6 +287,22 @@ theorem iIndepFun.cgf_sum {X : ι → Ω → ℝ}
+theorem mgf_congr_of_identDistrib
+    (X : Ω → ℝ) {Ω' : Type*} {m' : MeasurableSpace Ω'} {μ' : Measure Ω'} (X' : Ω' → ℝ)
+    (hident : IdentDistrib X X' μ μ') (t : ℝ) :
+    mgf X μ t = mgf X' μ' t := hident.comp (measurable_const_mul t).exp |>.integral_eq
```
Style:
```suggestion
  mgf X μ t = mgf X' μ' t := hident.comp (measurable_const_mul t).exp |>.integral_eq
```

RemyDegenne commented:
Thanks!
bors r+

RemyDegenne commented:
It looks like bors did not react? Let's try again.
bors r+
